### PR TITLE
Add Realm Live<T> data wrapper

### PR DESCRIPTION
**Pull request by peppy**
This is a proof of concept PR for review purposes. I still want to do a pass on it myself, and will leave some comments as I go.

- [ ] Depends on https://github.com/ppy/osu/pull/13618
- Also reverts temporary dumbing down of behaviour in above pull request to restore value change watching support.

### CODE REVIEW

File: osu.Game/Database/RealmExtensions.cs [hunk 1]
```diff
@@ -47,5 +48,31 @@ public static class RealmExtensions
...
+
+        /// <summary>
+        /// Wrap a managed instance of a realm object in a <see cref="Live{T}"/>.
+        /// </summary>
+        /// <param name="item">The item to wrap.</param>
+        /// <param name="contextFactory">A factory to retrieve realm contexts from.</param>
+        /// <typeparam name="T">The type of object.</typeparam>
+        /// <returns>A wrapped instance of the provided item.</returns>
+        public static Live<T> Wrap<T>(this T item, IRealmFactory contextFactory)
+            where T : RealmObject, IHasGuidPrimaryKey => new Live<T>(item, contextFactory);
```
Comment (bdach) [inline osu.Game/Database/RealmExtensions.cs:60]: As you even stated in the issue thread, I'd probably envision this being a method on `RealmContextFactory`, as the back-and-forth passing of `contextFactory` as written feels a bit off.

File: osu.Game/OsuGameBase.cs [hunk 2]
```diff
@@ -489,6 +498,9 @@ protected override void Dispose(bool isDisposing)
             LocalConfig?.Dispose();
 
             contextFactory.FlushConnections();
+
+            if (Host != null)
```
Comment (bdach) [inline osu.Game/OsuGameBase.cs:502]: I *believe* for extra robustness this check should be
```suggestion
            if (Host?.UpdateThread != null)
```

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/OsuGameBase.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/OsuGameBase.cs:502]: Seems fine, but this scenario can never happen (due to order of execution in `GameHost`).

File: osu.Game/Database/Live.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        public T Detach() => Get().Detach();
+
+        /// <summary>
+        /// Perform a write operation on this live object.
+        /// </summary>
+        /// <param name="perform">The action to perform.</param>
+        public void PerformUpdate(Action<T> perform)
+        {
+            using (contextFactory?.GetForWrite())
+                perform(Get());
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Database/Live.cs:86]: I was half tempted to move all the unused/untested methods in this class into a region mentioning such (or just deleting them completely).

I think instead, as part of this PR I will add proper test coverage for each of the methods / usages.

File: osu.Game/OsuGameBase.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/OsuGameBase.cs:502]: I've applied this in #13618.

File: osu.Game/Database/RealmExtensions.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Database/RealmExtensions.cs]: This helper method no longer exists.

**Comment by peppy** [comment]
I've made some pretty large refactors (and dumbing down) of `Live<T>`. @bdach would appreciate a re-review.

Will be adding test coverage for all the common scenarios I can foresee coming up.

Note that now retrieval of `Live<T>` must be done on the update thread. This is for simplicity, because with the refactoring (the class itself is now always responsible for the initial query) it's no longer possible to easily support the first fetch being done in an async context.

As an example, this would have previously worked:

```csharp
private void load(IRealmFactory realm)
{
    using (var usage = realm.GetForRead())
    {
          RealmObject retrieved = usage.Context.All<Things>().First();
          liveObject = new Live(retrieved);
    }
}

public override void LoadComplete()
{
    liveObject.Value.Text;
}
```

With the new structure, it's not easily possible to do the initial fetch from the async realm context. I did try moving the retrieval logic inside and always run it via `realm.GetForRead()`, but on top of it being quite a bit more complex to manage, the following scenario would not work without explicit `Live<T>` disposal support:

```csharp
private void load(IRealmFactory realm)
{
      var liveObject = new Live(() => usage.Context.All<Things>().First());
      
      // doing something with threaded retrieval
      liveObject.Value.Text;
}
```

as the internal *threaded* `RealmUsage` is not disposed.

---

I think we *will* want this functionality in the future so will look to add it back when that comes up, but just wanted to reduce the size and complexity of interactions as a starting point. Interested in whether you have any ideas on how to make it work in a sane way (making `Live<T>` disposable should be last resort IMO).

**Comment by peppy** [comment]
I've updated the usage guide to show how this should be consumed. I've also added two examples of issues with this current implementation, as inline comments on the usage guide. I don't think this is going to work without `IDisposable`. See what you think: https://github.com/ppy/osu/wiki/Realm-usage-rules

**Comment by bdach** [comment]
Bit hard to read through this diff given it's a few links away from `master` down the dependency chain but I do like the look of the new thinned-out `Live<T>`. I think while we don't have any access patterns worked out for realm that is probably the best way to go forward first, and the abstraction could be expanded if we keep running into the same operations on it in different contexts.

One thing I'm not sure of in the long run is eagerly re-retrieving all live objects after a flush - it may be better down the line to do that lazily to not take that penalty immediately. But that is a relatively simple change to make and it can be again made down the line when deemed really necessary.

**Comment by bdach** [comment]
Oh, and as for whether `Live<T>` should be disposable - I'm not necessarily sure what that would do, would it unregister the live from being refreshed in `RealmContextFactory`? If so then that sounds like a good idea, sure.

**Comment by peppy** [comment]
Without doing it immediately it loses all benefit, though.

In further discussion with @smoogipoo I think the best path forward may be to restart the game on thread mode change until we can find a better alternative (and throw this `Live<T>` thing away a second time).

**Comment by smoogipoo** [comment]
I'll hold off reviewing this until that's done and/or you get a response from Realm maintainers on the issue.

**Comment by peppy** [comment]
Tentatively closing this for now. Still leaning in the direction of removing immediate thread execution mode changes (osu! side) rather than adding this complexity layer. And before doing that, hoping there is another alternative.

**Pull request closed without merging**